### ISSUE TITLE
Export as runnable jar file

### ISSUE THREAD

**Issue by pgavigan** [NMAI-lab/SAVI#155]
We need to be able to build this project (with all dependencies, configuration files, asl files, image files, etc) into a single jar file that can be easily run.

**Comment by MikeVezina** [comment]
As it stands, I was able to generate a jar file with all files and dependencies. 

However, our code isn't set up to load files from within the jar file and so it expects the asl, image files, etc. to reside outside the jar file. 

I can fix this so that we can load files from within the jar file, but I'd like to hear thoughts on this first.

@pgavigan @ntartania @cruizm @gtrabes

**Comment by pgavigan** [comment]
> As it stands, I was able to generate a jar file with all files and dependencies.
> 
> However, our code isn't set up to load files from within the jar file and so it expects the asl, image files, etc. to reside outside the jar file.
> 
> I can fix this so that we can load files from within the jar file, but I'd like to hear thoughts on this first.
> 
> @pgavigan @ntartania @cruizm @gtrabes

Good job! What would you have to do to the code in order to make this work?

**Comment by MikeVezina** [comment]
One example would be changing from this:
```
String filePath = new File("").getAbsolutePath();
filePath = filePath + "/config.cfg";
System.out.println(filePath);
File inFile = new File(filePath);
in = new FileInputStream(inFile);
modelProps.load(in);
```

To something along the lines of:
```
File file = new File(getClass().getResource("/config.cfg").toURI());
in = new FileInputStream(file);
modelProps.load(in);
```

This would have to be done for all files that are loaded (.asl, images, etc.)

**Comment by ntartania** [comment]
I looked into this, and the main difficulty is that the images are loaded by internal Processing code, which is not written in such a way that we can switch from the file name / path to getClass().getResource().
There *could* be a way out by subclassing /rewriting the methods that load images, but it's a bit involved and perhaps not worth the trouble. I think it would be ok to release the code as a zip or something, that includes a runnable jar and separate image/ etc. files.
There's also a but related to the jogl and gluegen jar files, which don't seem to work properly once experted to a runnable jar. The solution is to replace them by another library called jogamp-fat.jar. The nice thing is that it reduces the dependencies from about 6 to 3 (4 on mac).

**Comment by pgavigan** [comment]
I think @ntartania is suggesting a reasonable approach (at least for now given the state of the project in the grand scheme of things).

**Comment by pgavigan** [comment]
This should include documentation for how to generate this jar as well as how to build the project. This should include how to build the project, etc.

**Comment by MikeVezina** [comment]
@pgavigan @ntartania @gtrabes @cruizm Can everyone try running both JARs in the attachment to make sure everything runs as expected?

[Runnable JARs.zip](https://github.com/NMAI-lab/SAVI/files/3163898/Runnable.JARs.zip)

**Comment by pgavigan** [comment]
> @pgavigan @ntartania @gtrabes @cruizm Can everyone try running both JARs in the attachment to make sure everything runs as expected?
> 
> [Runnable JARs.zip](https://github.com/NMAI-lab/SAVI/files/3163898/Runnable.JARs.zip)

Worked well on my machine. Good job!

**Comment by ntartania** [comment]
Didn't work for me...

java.lang.UnsatisfiedLinkError: Can't load library: [where I unzipped the
zip]/SAVI/natives/macosx-universal//gluegen-rt

at java.lang.ClassLoader.loadLibrary(ClassLoader.java:1827)

at java.lang.Runtime.load0(Runtime.java:809)

at java.lang.System.load(System.java:1086)

at
com.jogamp.common.jvm.JNILibLoaderBase.loadLibraryInternal(JNILibLoaderBase.java:624)

at
com.jogamp.common.jvm.JNILibLoaderBase.access$000(JNILibLoaderBase.java:63)

at
com.jogamp.common.jvm.JNILibLoaderBase$DefaultAction.loadLibrary(JNILibLoaderBase.java:106)

at
com.jogamp.common.jvm.JNILibLoaderBase.loadLibrary(JNILibLoaderBase.java:487)

at
com.jogamp.common.os.DynamicLibraryBundle$GlueJNILibLoader.loadLibrary(DynamicLibraryBundle.java:421)

at com.jogamp.common.os.Platform$1.run(Platform.java:317)

at java.security.AccessController.doPrivileged(Native Method)

at com.jogamp.common.os.Platform.<clinit>(Platform.java:287)

at
com.jogamp.nativewindow.NativeWindowFactory$2.run(NativeWindowFactory.java:185)

at java.security.AccessController.doPrivileged(Native Method)

at
com.jogamp.nativewindow.NativeWindowFactory.<clinit>(NativeWindowFactory.java:182)

at com.jogamp.newt.NewtFactory$1.run(NewtFactory.java:68)

at java.security.AccessController.doPrivileged(Native Method)

at com.jogamp.newt.NewtFactory.<clinit>(NewtFactory.java:65)

at processing.opengl.PSurfaceJOGL.initIcons(PSurfaceJOGL.java:498)

at processing.opengl.PSurfaceJOGL.initFrame(PSurfaceJOGL.java:134)

at processing.core.PApplet.initSurface(PApplet.java:10924)

at processing.core.PApplet.runSketch(PApplet.java:10811)

at processing.core.PApplet.main(PApplet.java:10511)

at savi.jason_processing.SAVIWorld_model.main(SAVIWorld_model.java:84)

**Comment by pgavigan** [comment]
I'm wondering if this is a result of building the jar on Windows and then trying to run it on a Mac... could there be some internal configuration that is being used that makes the jar that @MikeVezina is building not compatible with the computer that @ntartania is using. I'm curious, if @ntartania were to make a jar, would I, or @MikeVezina be able to run it (or would we have similar issues)?

If we were to find this to be the case, perhaps we can resolve this issue by providing a proceedure for building the jar (using some IDE) and also provide jar files on our website for Windows and Mac OS. I think this would cover any requirements we may have for delivery and avoid the need for complicated debugging of the jar file.

Thoughts?

**Comment by MikeVezina** [comment]
I foolishly created the Jar without any of the Mac/Linux dependencies. Try this one @ntartania @pgavigan

I tested this out on my mac (& windows machine) and both worked fine. Has not been tested on linux.

[Runnable JARs.zip](https://github.com/NMAI-lab/SAVI/files/3169315/Runnable.JARs.zip)

**Comment by pgavigan** [comment]
> I foolishly created the Jar without any of the Mac/Linux dependencies. Try this one @ntartania @pgavigan
> 
> I tested this out on my mac (& windows machine) and both worked fine. Has not been tested on linux.
> 
> [Runnable JARs.zip](https://github.com/NMAI-lab/SAVI/files/3169315/Runnable.JARs.zip)
> 
> I've been using IntelliJ to build the Jars and it seems to do a pretty good job at managing & organizing which dependencies get extracted to the jar file. Although, im sure you could do the same in eclipse.

Worked for me.

**Comment by ntartania** [comment]
It runs for me too!
However, this doesn't seem to be the latest code. The perception circles
are off.

**Comment by MikeVezina** [comment]
Glad to hear!

I have updated them to include the latest commits! If I get some time this week I may look into a makefile to make this easier for us.

[Updated JARs.zip](https://github.com/NMAI-lab/SAVI/files/3173906/Updated.JARs.zip)

Side note: the GUI for the command station (on Mac Only) looks a bit wonky to me. Every action is still accessible but the frame components overlap a little bit. Not a big issue by any means.

**Comment by gtrabes** [comment]
Looks and works good for me too (on windows), 
Good job!

**Comment by ntartania** [comment]
The interface is wonky because the text fields are too wide (the text
fields could work with about 50 char).

**Comment by ntartania** [comment]
Finally I'd like to add a request to clean up the output: perhaps put a verbose flag to deactivate them entirely. 
I realize you don't get them if you double click on the jar but if you launch it from the command line you get the very busy console and I think that's not great.

**Comment by pgavigan** [comment]
> Finally I'd like to add a request to clean up the output: perhaps put a verbose flag to deactivate them entirely.
> I realize you don't get them if you double click on the jar but if you launch it from the command line you get the very busy console and I think that's not great.
> (Not this request is not specifically aimed at one person...)

Should this be a separate issue?

**Comment by pgavigan** [comment]
Resolved by PR #166.
